### DISCUSSION TITLE
Nature and number of IP addresses an Application is accessible to its clients

### DISCUSSION THREAD

**Discussion by gunjald** [camaraproject/EdgeCloud#264] [General]
The App LCM API does not explicitly state the IP scheme for the instances of the an application in terms of the instance availability to its clients. There could be applications which will have one or more components say N and then there could be even multiple instances of the application say M. That means in theory there could be NxM endpoints that application can have. 
Also in TI API the API consumer may need to use information like IP address of the application (or app instance) to be used to for the influence and based on the application design it may use different IPs to refer to different traffic flows.

I think we need to define the IP address model associated to the application and its nature like public or private or mix of both in different conditions so as to convey it correctly to the API users and set the right expectation.

**Comment by gainsley** [comment]
Please also see related issue #262 which recommends to have a single componentspec for an AppManifest in the [Edge-Application-Management spec](https://github.com/camaraproject/EdgeCloud/blob/main/code/API_definitions/Edge-Application-Management.yaml).

I'm not sure why the number of instances of an application matters for TI, as a device will choose through discovery a single application instance to connect to, right?

By public/private ports I assume you are referring to the port visibility:
```
                    visibilityType:
                      description: |
                        Defines whether the interface is exposed
                        to outer world or not i.e., external, or internal.
                        If this is set to "external", then it is  exposed
                        to external applications otherwise it is exposed
                        internally to edge application components within
                        edge cloud. When exposed to external world,
                        an external dynamic port is assigned for UC traffic
                        and mapped to the extPort
                      type: string
                      enum:
                        - VISIBILITY_EXTERNAL
                        - VISIBILITY_INTERNAL
```

It has been my understanding that only `external` ports are accessible by a device, and `internal` are used for appinstance-to-appinstance or communication internal to edge cloud sites. We should clarify the accessibility in the context of TI.

**Reply by gunjald** [reply]
Yes device may chose a single instance of an app but then as the app itself may contain multiple components so each component can have its own IP possibly unless we define what is possible. In the App Endpoint discovery API we are still having an open point to have one or more endpoints which can also result into potentially multiple IP addresses for a single instance of an application.

The key point in my opinion is clearly define what is the model to be followed, how to interpret the current structure of endpoints to make the API behavior explicit. Also, as various APIs are correlated in some manner it is really required to make that clarity documented.

**Reply by gainsley** [reply]
So then it's just N instead of NxM. Here we're assuming each componentspec only has a single IP.
